src/crimson/os/seastore/onode_manager/staged-fltree/fltree_onode_manager.cc

```diff
@@ -141,21 +143,28 @@ FLTreeOnodeManager::list_onodes_ret FLTreeOnodeManager::list_onodes(
       return trans_intr::repeat(
           [this, &trans, end, &to_list, &current_cursor, &ret] ()
           -> eagain_ifuture<seastar::stop_iteration> {
+	LOG_PREFIX(FLTreeOnodeManager::list_onodes);
         if (current_cursor.is_end()) {
+	  DEBUGT("reached the onode tree end", trans);
           std::get<1>(ret) = ghobject_t::get_max();
           return seastar::make_ready_future<seastar::stop_iteration>(
             seastar::stop_iteration::yes);
         } else if (current_cursor.get_ghobj() >= end) {
-          std::get<1>(ret) = end;
```

Comment:
Um, I was thinking both `the listing has reached the very last` and `the listing has listed all the items within the input range` means that there's nothing more to be listed, so they can be treated the same.

Reply:
Backfill scanning is not the only user, I think the interface by itself should be able to distinguish these cases to be generic.

It should also be easy to change the user side to check if std::get<1>(ret) is either max or end for this purpose (i.e. std::get<1>(ret) >= end), right?

Reply:
> I think the upper layer users don't care whether the list has reached the end of the tree or not as it's the internal state of the btree-based onode manager, am I right?

I'm not sure whether all the upper layer users don't care about it currently. Probably @athanatos @rzarzynski can add more to this part. Even so, I think these cases are not internal and can/should be distinguished by the interface.

Reply:
> Backfill scanning is not the only user, I think the interface by itself should be able to distinguish these cases to be generic.
> 
> It should also be easy to change the user side to check if std::get<1>(ret) is either max or end for this purpose (i.e. std::get<1>(ret) >= end), right?

It's not that easy, because we're reusing a lot of backfill related data structures from the classic osd which don't encapsulate the list ranges and rely on this `is_max()` interface to judge whether the scan has reached the end. And I think that is for a reason, @athanatos what do you think?

Reply:
Another issue is that the original `ObjectStore::collection_list(c, start, end, max, ls, next)` https://github.com/ceph/ceph/blob/main/src/os/ObjectStore.h#L678 doesn't seem to provide sufficient comments about how `ghobject_t::get_max()` should be returned as `next`, which causes the confusion here.

Reply:
Yeah, IIUC, the semantic of `ObjectStore::collection_list(c, start, end, max, ls, next)` is:
1. list all objects in the collection `c` within the range `start` and `end`
2. the list may have at most `max` items
3. if the list exhausted all the objects within the range before reaching the end, return `ghobject_t::max()`

When we do backfill scanning, the `end` passed to `ObjectStore::collection_list()` is `ghobject_t::max()` as it doesn't what the last object is, so `SeaStore::list_objects()` needs to return `ghobject_t::max()` when it has listed the last object in the collection `c`. Note that `the last object in the collection c` doesn't necessarily mean `the last onode in the onode tree`.

This is the reason why this PR is submitted.

Sorry, I didn't clarify this at the beginning.

Reply:
If I'm correct in the [above comment](https://github.com/ceph/ceph/pull/53062#discussion_r1351269373), I think we shouldn't change `OnodeManager` indeed and should return to the original modification in which all the modifications are done in `SeaStore::list_objects()`, am I right?

Reply:
I see, looks the original `SeaStore::list_objects()` violates https://github.com/ceph/ceph/pull/53062#discussion_r1338333338 about the 3rd question that the input `end` is `ghobject_t::max()` and the listing has listed all the items within the input range with ch, it fails to return `std::get<1>(ret)` as `end/ghobject_t::max()`.

Reply:
Yeah, I'll see what I can do.

Reply:
> I see, looks the original `SeaStore::list_objects()` violates [#53062 (comment)](https://github.com/ceph/ceph/pull/53062#discussion_r1338333338) about the 3rd question that the input `end` is `ghobject_t::max()` and the listing has listed all the items within the input range with ch, it fails to return `std::get<1>(ret)` as `end/ghobject_t::max()`.

Please take a look again, thanks:-)